### PR TITLE
Add MultiGroup support

### PR DESCRIPTION
**Pull request by SaeedYasin**
This is a proposed implementation of the enhancement suggested in #1165

### CODE REVIEW

**Comment by SaeedYasin** [comment]
@basvodde @jacob-keller does this PR look okay? If possible please provide some feedback, it's been a long time. Thank you.

**Comment by basvodde** [comment]
Sorry @SaeedYasin , I've been super busy and thus usually do not spend enough time on PRs that are large and need some focus.

As it is, it is unlikely to be included in the core of CppUTest. The implementation contains quite a lot of macro code, which we tend to avoid. I've written a lot of tests in CppUTest and never really needed this feature as I can just have multiple tests and call a setup in the test body instead of in a setup.

My initial reaction is to just decline it (sorry) as too complicated for a rare specific scenario.

**Comment by jacob-keller** [comment]
I think there are two cases here:

1st) setup specific to a test. For this, you can define a function with the parameters of the test in the test class and then call it as the first step of the test with the specific test parameters.

2nd) multiple groups of tests where they all share underlying setup and then each group needs its own setup.

For this, my teams tend to use TEST_BASE and TEST_GROUP_BASE. Additionally, we found a convenient "hack" that allows us to also use inheritance on test classes not defined as TEST_BASE:

We define a special macro, TGN:

`#define TGN(group) TEST_GROUP_##CppUTestGroup##group`

this lets us write things like

```
TEST_GROUP_BASE(new_group, TGN(existing_group)) {
  TEST_SETUP {
    TGN(existing_group)::setup()

    // extended setup
  }
}
```

Basically using the TGN macro lets us extend an existing non-base class easily.


However, looking closer, that doesn't fully resolve your problem. I think you should provide more detail and examples.

Your goal is to cover the situation where you have tests that you want to repeat with many different sets of conditions, without having multiple tests defined for each one.

That can definitely be helpful, but I wonder if there is a better way to implement and explain the goal.

Essentially you want to parameterize a test group to expand it based on initial conditions to run many copies with different setups. I suspect there has to be a good way to do what you want. I know other test frame works have somethings similar to this...

**Comment by SaeedYasin** [comment]
@basvodde completely understandable, the reason for the macro based implementation was, that it makes it possible to achieve this functionality with zero risk to the existing code. But anyway, thank you for the feedback.

@jacob-keller yes the goal is to "have tests that you want to repeat with many different sets of conditions, without having multiple tests defined for each one". This removes test code duplication and makes it easier to understand and maintain tests, when the number of tests is high.

THE TGN macro and TEST_GROUP_BASE techniques look very interesting and definitely useful in certain cases. Thank you for the guidance and feedback.
